HSP2tools/readWDM.py

```diff
@@ -87,60 +113,56 @@ def readWDM(wdmfile, hdffile, jupyterlab=True):
                 ptr = iarray[index + i + 1] - 1 + index
-                if id not in attrinfo:
+                if id not in self.attrinfo:
                     # print('PROGRAM ERROR: ATTRIBUTE INDEX not found', id, 'Attribute pointer', iarray[index + i+1])
                     continue
 
-                name, atype, length = attrinfo[id]
+                name, atype, length = self.attrinfo[id]
                 if atype == 'I':
                     dattr[name] = iarray[ptr]
                 elif atype == 'R':
                     dattr[name] = farray[ptr]
                 else:
-                    dattr[name] = ''.join([itostr(iarray[k]) for k in range(ptr, ptr + length//4)]).strip()
+                    dattr[name] = ''.join([WDMReader._inttostr(iarray[k]) for k in range(ptr, ptr + length//4)]).strip()
 
             # Get timeseries timebase data
-            records = []
+            records = [] 
+            offsets = []
             for i in range(pdat+1, pdatv-1):
                 a = iarray[index+i]
                 if a != 0:
-                    records.append(splitposition(a))
+                    record, offset = WDMReader._splitposition(a)
+                    records.append(record)
+                    offsets.append(offset)
             if len(records) == 0:
-                continue   # WDM preallocation, but nothing saved here yet
-
-            srec, soffset = records[0]
-            start = splitdate(iarray[srec*512 + soffset])
-
-            sprec, spoffset = splitposition(frepos)
-            finalindex = sprec * 512 + spoffset
+                continue   
 
             # calculate number of data points in each group, tindex is final index for storage
             tgroup = dattr['TGROUP']
             tstep  = dattr['TSSTEP']
             tcode  = dattr['TCODE']
-            cindex = pd.date_range(start=start, periods=len(records)+1, freq=freq[tgroup])
-            tindex = pd.date_range(start=start, end=cindex[-1], freq=str(tstep) + freq[tcode])
-            counts = np.diff(np.searchsorted(tindex, cindex))
-
-            ## Get timeseries data
-            floats = np.zeros(sum(counts),  dtype=np.float32)
-            findex = 0
-            for (rec,offset),count in zip(records, counts):
-                findex = getfloats(iarray, farray, floats, findex, rec, offset, count, finalindex, tcode, tstep)
-
-            ## Write to HDF5 file
-            series = pd.Series(floats[:findex], index=tindex[:findex])
+
+            records = np.asarray(records)
+            offsets = np.asarray(offsets)
+            # process groups in the time series
+            dates, values = self._process_groups(iarray, farray, records, offsets, tgroup)
+            dates = np.array(dates)
+            dates_converted = date_convert(dates, date_epoch, dt_year, dt_month, dt_day, dt_hour, dt_minute, dt_second)
+            series = pd.Series(values, index=dates_converted)
+            
+            self.datasets[dsn] = series
```

Comment:
I'm concerned that storing all of the DSN series in a dictionary will be too memory intensive for large WDM files. See my comment about about splitting this function into some smaller pieces. I think we should have a function which processes a single DSN and return this series which is immediately written to a file (or better yet passed to the eventual IO abstraction class for output into the desired output format).